client/lib/url/format.ts

```diff
@@ -0,0 +1,75 @@
+/**
+ * External dependencies
+ */
+import { URL as URLString } from 'types';
+import { Falsy } from 'utility-types';
+
+/**
+ * Internal dependencies
+ */
+import { determineUrlType, URL_TYPE } from './url-type';
+
+const BASE_URL = 'http://__domain__.invalid/';
+
+/**
+ * Format a URL as a particular type.
+ * Does not support formatting invalid or path-relative URLs.
+ *
+ * @param url The URL to format.
+ * @param urlType The URL type into which to format. If not provided, defaults to the same type as
+ *                `url`, which effectively results in the URL being normalized.
+ *
+ * @returns The formatted URL.
+ */
+export default function format(
+	url: URLString | URL | Falsy,
```

Comment:
Here, and Looking back at https://github.com/Automattic/wp-calypso/pull/37030:

https://github.com/Automattic/wp-calypso/blob/f53864b6e5ce396c851c273813e1fcae6fff0bc8/client/lib/url/url-type.ts#L36

Why do we want to accept `Falsy` values in these new implementations?

Reply:
The idea was to match other `lib/url` methods which also take `Falsy`. Do you think it would be best not to accept it, while keeping the current implementation (to handle `Falsy` values passed in from untyped code)?

Reply:
I've been thinking about this for several days, and there's nuance here that depends on what exactly where doing and where as we work on a gradually typed application.

That said, my current thinking for new code is that we should be working with narrow types rather than more permissive one. This should make the API clear and puts the onus on the consumer to use the functionality properly.

When refactoring existing code from JavaScript to TypeScript that's already in use it's not as clear to me, but that's not the case here and is something we can sort out in another moment 🙂

Reply:
That's a very good set of points. There's no reason why we should accept values we can't handle just because other functions in the same library currently do that. If falsy values don't make sense in this context, they shouldn't be part of the accepted types.

As discussed, I decided to do a runtime check for the types too, since this method will see significant usage in JS contexts. The meaningful error messages that come with those checks should make debugging easier.